### ISSUE TITLE
Support of GCC Builtins

### ISSUE THREAD

**Issue by mrigger** [AbsInt/CompCert#243]
We are currently studying the usage of [GCC builtins](https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html) by GitHub projects and how well various tools support them. To that end, we are also developing a test suite for the most frequently used machine-independent GCC builtins (available at https://github.com/gcc-builtins/tests/tree/master/test-cases). We'd be glad to contribute the test suite (or parts of it) to the project, if you are interested.

We tested CompCert 3.3 with this test suite and found that currently only 9 builtins are supported (`__builtin_ctz`, `__builtin_bswap64`, `__builtin_ctzll`, `__builtin_bswap32`, `__builtin_offsetof`, `__builtin_clz`, `__builtin_fabs`, `__builtin_clzll`, and `__builtin_bswap16`). Additionally, it seems that some atomic builtins are supported  partially (see https://github.com/AbsInt/CompCert/pull/54).

Furthermore, it appears that our test cases uncovered two erroneous implementations of builtins (`__builtin_clzl` and `__builtin_ctzl`). I attached the reduced versions of the test cases (see below). It seems that instead of the `unsigned long` input, an `int` input is expected in both `__builtin_clzl` and `__builtin_ctzl`.

```c
#include <assert.h>

int main() {
  unsigned long value = (unsigned long) -1 >> 3;
  assert(__builtin_clzl(value) == 3);
}
```

```c
#include <assert.h>

int main() {
  unsigned long value = 0x10000000000LL;
  assert(__builtin_ctzl(value) == 40);
}
```

**Comment by mrigger** [comment]
I've only tested x86 so far. I sent you the (preliminary) test harness per email, if you want to test for other architectures. It would be great if you could tell me if you found other bugs with it.

**Comment by m-schmidt** [comment]
There is one problem: `__builtin_clzl` is designed for a `long` type with 32bit, but `long` on x86_64 is 64bits wide. If you use `__builtin_clzll` and `__builtin_ctzll` both assertion hold.

**Comment by mrigger** [comment]
I've already speculated that this would be the case. Yes, the test cases for the other integer types of `clz` and `ctz` pass.

**Comment by m-schmidt** [comment]
The `xor` is correct, since it flips the lower bits (1 complement) and that converts the bit position of the scan instruction into a bit count.

So the bug is, that on 64bit architectures the `long`-versions of the builtins should expand the code of the `long long` version. Therefore 64bit RISC-V may also be affected.

**Comment by xavierleroy** [comment]
Fixed in commit abf674c.  Thanks @mrigger for the report and @m-schmidt for the fix.

The test suite for GCC builtins looks very interesting.  CompCert's formal verification doesn't cover builtin functions, so we need more testing here.
